src/merkle/smt.rs

```diff
@@ -0,0 +1,217 @@
+use winter_math::StarkField;
+
+use crate::{
+    hash::rpo::{Rpo256, RpoDigest},
+    Word,
+};
+
+use super::{MerkleError, MerklePath, NodeIndex, Vec};
+
+// CONSTANTS
+// ================================================================================================
+
+/// Minimum supported depth.
+pub const SMT_MIN_DEPTH: u8 = 1;
+
+/// Maximum supported depth.
+pub const SMT_MAX_DEPTH: u8 = 64;
+
+// SPARSE MERKLE TREE
+// ================================================================================================
+
+/// An abstract description of a sparse Merkle tree.
+///
+/// A sparse Merkle tree is a key-value map which also supports proving that a given value is indeed
+/// stored at a given key in the tree. It is viewed as always being fully populated. If a leaf's
+/// value was not explicitly set, then its value is the default value. Typically, the vast majority
+/// of leaves will store the default value (hence it is "sparse"), and therefore the internal
+/// representation of the tree will only keep track of the leaves that have a different value from
+/// the default.
+///
+/// All leaves sit at the same depth. The deeper the tree, the more leaves it has; but also the
+/// longer its proofs are - of exactly `log(depth)` size. A tree cannot have depth 0, since such a
+/// tree is just a single value, and is probably a programming mistake.
+///
+/// Every key value maps to one leaf. If there are as many keys as there are leaves, then
+/// [Self::Leaf] should be the same type as [Self::Value], as is the case with
+/// [crate::merkle::SimpleSmt]. However, if there are more keys than leaves, then [`Self::Leaf`]
+/// must accomodate all keys that map to the same leaf.
+///
+/// [SparseMerkleTree] currently doesn't support optimizations that compress Merkle proofs.
+pub trait SparseMerkleTree<const DEPTH: u8> {
+    /// The type for a key, which must be convertible into a `u64` infaillibly
+    type Key: Into<LeafIndex<DEPTH>> + Clone;
+    /// The type for a value
+    type Value: Clone + Default + PartialEq;
```

Comment:
Instead of `Default`, we might want to have an abstract method `empty_value()`, in the case where someone wants to build a tree where the default value is different from `Default` (e.g. where the default `Word` is not the empty word).

From a UX perspective too, it means that the implementor pays attention to the implementation of `Default` for the `Value` type, and makes sure that it makes sense in the context of a sparse merkle tree